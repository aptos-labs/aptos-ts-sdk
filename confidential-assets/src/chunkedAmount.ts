import { AnyNumber } from "@aptos-labs/ts-sdk";

/**
 * Number of chunks for confidential balance
 */
const CHUNKS_COUNT = 8;

/**
 * Max bits of amount in a chunk for normalized confidential balance
 */
const CHUNK_BITS = 16;

export class ChunkedAmount {
  amount: bigint;

  amountChunks: bigint[];

  chunksCount = CHUNKS_COUNT;

  chunkBits = CHUNK_BITS;

  static CHUNKS_COUNT = CHUNKS_COUNT;

  static CHUNK_BITS = CHUNK_BITS;

  static CHUNK_BITS_BIG_INT = BigInt(ChunkedAmount.CHUNK_BITS);

  static CHUNKS_COUNT_HALF = ChunkedAmount.CHUNKS_COUNT / 2;

  constructor(args: { amount?: AnyNumber; amountChunks: bigint[]; chunksCount?: number; chunkBits?: number }) {
    this.amount = args.amount ? BigInt(args.amount) : ChunkedAmount.chunksToAmount(args.amountChunks);
    this.amountChunks = args.amountChunks;

    if (args.chunksCount) {
      this.chunksCount = args.chunksCount;
    }

    if (args.chunkBits) {
      this.chunkBits = args.chunkBits;
    }
  }

  /**
   * Returns the original amount from the provided chunks,
   * where each chunk is represented as a 32 bit number
   *
   * amount = a0 + a1 * (2 ** 32) + a2 * (2 ** 64) ... a_i * (2 ** 32 * i)
   */
  static chunksToAmount(chunks: bigint[], chunkBits = ChunkedAmount.CHUNK_BITS): bigint {
    const chunkBitsBi = BigInt(chunkBits);

    return chunks.reduce((acc, chunk, i) => acc + chunk * 2n ** (chunkBitsBi * BigInt(i)), 0n);
  }

  /**
   * Splits a given bigint amount into an array of smaller "chunk" values (also as bigints).
   *
   * @param amount - The original amount as a bigint.
   * @param chunksCount - The number of chunks to split the amount into (default is ConfidentialAmount.CHUNKS_COUNT).
   * @param chunkBits - The number of bits each chunk should contain (default is ConfidentialAmount.CHUNK_BITS).
   * @returns An array of bigints, where each element represents a segment of the original amount.
   */
  static amountToChunks(
    amount: AnyNumber,
    chunksCount = ChunkedAmount.CHUNKS_COUNT,
    chunkBits = ChunkedAmount.CHUNK_BITS,
  ): bigint[] {
    // Initialize an empty array that will hold the chunked values.
    const chunks: bigint[] = [];

    // Convert chunkBits to a bigint so it can be used in bitwise operations with `amount`.
    const chunkBitsBi = BigInt(chunkBits);

    // Loop over the total number of chunks we want.
    for (let i = 0; i < chunksCount; i++) {
      /**
       * 1. Shift the amount right by (chunkBitsBi * i) bits to move the desired chunk
       *    into the lowest bits of the number.
       *
       * 2. Use a bitmask ( (1n << chunkBitsBi) - 1n ) to extract only those `chunkBits`
       *    bits. This mask is effectively a number with `chunkBits` 1s in binary.
       */
      const chunk = (BigInt(amount) >> (chunkBitsBi * BigInt(i))) & ((1n << chunkBitsBi) - 1n);

      // Add this extracted chunk to the chunks array.
      chunks.push(chunk);
    }

    // Return the array of chunked values.
    return chunks;
  }

  static fromAmount(
    amount: AnyNumber,
    opts?: {
      chunksCount?: number;
      chunkBits?: number;
    },
  ): ChunkedAmount {
    const amountChunks = ChunkedAmount.amountToChunks(amount, opts?.chunksCount, opts?.chunkBits);
    return new ChunkedAmount({ amount, amountChunks, ...opts });
  }

  static fromChunks(chunks: bigint[]): ChunkedAmount {
    const amount = ChunkedAmount.chunksToAmount(chunks);

    return new ChunkedAmount({ amount, amountChunks: chunks });
  }
}
