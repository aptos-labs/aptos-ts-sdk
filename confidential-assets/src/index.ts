export * from './confidentialAmount'
export * from './confidentialAsset'
export * from './confidentialKeyRotation'
export * from './confidentialNormalization'
export * from './confidentialTransfer'
export * from './confidentialWithdraw'
export * from './consts'
export * from './helpers'
export * from './rangeProof'
export * from './twistedEd25519'
export * from './twistedElGamal'
export * from './utils'
