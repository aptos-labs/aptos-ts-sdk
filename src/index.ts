// Copyright Â© Aptos Foundation
// SPDX-License-Identifier: Apache-2.0

export * from "./api";
export * from "./bcs";
export * from "./client";
export * from "./core";
export * from "./transactions";
export * from "./transactions/management";
export * from "./types";
export * from "./utils";
