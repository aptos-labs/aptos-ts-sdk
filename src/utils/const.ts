// Copyright Â© Aptos Foundation
// SPDX-License-Identifier: Apache-2.0

import { Network } from "./apiEndpoints";

/**
 * Type of API endpoint for request routing
 */
export enum AptosApiType {
  FULLNODE,
  INDEXER,
  FAUCET,
}

export const DEFAULT_NETWORK = Network.DEVNET;

/**
 * The default max gas amount when none is given.
 *
 * This is the maximum number of gas units that will be used by a transaction before being rejected.
 *
 * Note that max gas amount varies based on the transaction.  A larger transaction will go over this
 * default gas amount, and the value will need to be changed for the specific transaction.
 */
export const DEFAULT_MAX_GAS_AMOUNT = 200000;

/**
 * The default transaction expiration seconds from now.
 *
 * This time is how long until the blockchain nodes will reject the transaction.
 *
 * Note that the transaction expiration time varies based on network connection and network load.  It may need to be
 * increased for the transaction to be processed.
 */
export const DEFAULT_TXN_EXP_SEC_FROM_NOW = 20;

/**
 * The default number of seconds to wait for a transaction to be processed.
 *
 * This time is the amount of time that the SDK will wait for a transaction to be processed when waiting for
 * the results of the transaction.  It may take longer based on network connection and network load.
 */
export const DEFAULT_TXN_TIMEOUT_SEC = 20;

/**
 * The default gas currency for the network.
 */
export const APTOS_COIN = "0x1::aptos_coin::AptosCoin";

export const RAW_TRANSACTION_SALT = "APTOS::RawTransaction";
export const RAW_TRANSACTION_WITH_DATA_SALT = "APTOS::RawTransactionWithData";

/**
 * The list of supported Processor types for our indexer api.
 *
 * These can be found from the processor_status table in the indexer database.
 * {@link https://cloud.hasura.io/public/graphiql?endpoint=https://indexer.mainnet.aptoslabs.com/v1/graphql}
 */
export enum ProcessorType {
  ACCOUNT_TRANSACTION_PROCESSOR = "account_transactions_processor",
  DEFAULT = "default_processor",
  EVENTS_PROCESSOR = "events_processor",
  // Fungible asset processor also handles coins
  FUNGIBLE_ASSET_PROCESSOR = "fungible_asset_processor",
  STAKE_PROCESSOR = "stake_processor",
  // Token V2 processor replaces Token processor (not only for digital assets)
  TOKEN_V2_PROCESSOR = "token_v2_processor",
  USER_TRANSACTION_PROCESSOR = "user_transaction_processor",
}
