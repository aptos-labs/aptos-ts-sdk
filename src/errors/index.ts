import { AptosApiType } from "../utils/const";
import { getErrorMessage } from "../utils/helpers";
import { AptosRequest, AptosResponse } from "../types";

export enum KeylessErrorCategory {
  API_ERROR,
  EXTERNAL_API_ERROR,
  SESSION_EXPIRED,
  INVALID_STATE,
  UNKNOWN,
}

export enum KeylessErrorResolutionTip {
  REAUTHENTICATE = "Re-authentiate to continue using your keyless account",
  // eslint-disable-next-line max-len
  REAUTHENTICATE_UNSURE = "Try re-authentiating. If the error persists join the telegram group at https://t.me/+h5CN-W35yUFiYzkx for further support",
  UPDATE_REQUEST_PARAMS = "Update the invalid request parameters and reauthenticate.",
  // eslint-disable-next-line max-len
  RATE_LIMIT_EXCEEDED = "Cache the keyless account and reuse it to avoid making too many requests.  Keyless accounts are valid until either the EphemeralKeyPair expires, when the JWK is rotated, or when the proof verifying key is changed, whichever comes soonest.",
  // eslint-disable-next-line max-len
  SERVER_ERROR = "Try again later.  See aptosApiError error for more context. For additional support join the telegram group at https://t.me/+h5CN-W35yUFiYzkx",
  // eslint-disable-next-line max-len
  CALL_PRECHECK = "Call `await account.checkKeylessAccountValidity()` to wait for asyncronous changes and check for account validity before signing or serializing.",
  REINSTANTIATE = "Try instantiating the account again.  Avoid manipulating the account object directly",
  JOIN_SUPPORT_GROUP = "For support join the telegram group at https://t.me/+h5CN-W35yUFiYzkx",
  UNKNOWN = "Error unknown. For support join the telegram group at https://t.me/+h5CN-W35yUFiYzkx",
}

export enum KeylessErrorType {
  EPHEMERAL_KEY_PAIR_EXPIRED,

  PROOF_NOT_FOUND,

  ASYNC_PROOF_FETCH_FAILED,

  INVALID_PROOF_VERIFICATION_FAILED,

  INVALID_PROOF_VERIFICATION_KEY_NOT_FOUND,

  INVALID_JWT_SIG,

  INVALID_JWT_JWK_NOT_FOUND,

  INVALID_JWT_ISS_NOT_RECOGNIZED,

  INVALID_JWT_FEDERATED_ISS_NOT_SUPPORTED,

  INVALID_TW_SIG_VERIFICATION_FAILED,

  INVALID_TW_SIG_PUBLIC_KEY_NOT_FOUND,

  INVALID_EXPIRY_HORIZON,

  JWT_PARSING_ERROR,

  JWK_FETCH_FAILED,

  JWK_FETCH_FAILED_FEDERATED,

  RATE_LIMIT_EXCEEDED,

  PEPPER_SERVICE_INTERNAL_ERROR,

  PEPPER_SERVICE_BAD_REQUEST,

  PEPPER_SERVICE_OTHER,

  PROVER_SERVICE_INTERNAL_ERROR,

  PROVER_SERVICE_BAD_REQUEST,

  PROVER_SERVICE_OTHER,

  FULL_NODE_CONFIG_LOOKUP_ERROR,

  FULL_NODE_VERIFICATION_KEY_LOOKUP_ERROR,

  FULL_NODE_JWKS_LOOKUP_ERROR,

  FULL_NODE_OTHER,

  UNKNOWN,
}

const KeylessErrors: { [key in KeylessErrorType]: [string, KeylessErrorCategory, KeylessErrorResolutionTip] } = {
  [KeylessErrorType.EPHEMERAL_KEY_PAIR_EXPIRED]: [
    "The ephemeral keypair has expired.",
    KeylessErrorCategory.SESSION_EXPIRED,
    KeylessErrorResolutionTip.REAUTHENTICATE,
  ],
  [KeylessErrorType.PROOF_NOT_FOUND]: [
    "The required proof could not be found.",
    KeylessErrorCategory.INVALID_STATE,
    KeylessErrorResolutionTip.CALL_PRECHECK,
  ],
  [KeylessErrorType.ASYNC_PROOF_FETCH_FAILED]: [
    "The required proof failed to fetch.",
    KeylessErrorCategory.INVALID_STATE,
    KeylessErrorResolutionTip.REAUTHENTICATE_UNSURE,
  ],
  [KeylessErrorType.INVALID_PROOF_VERIFICATION_FAILED]: [
    "The provided proof is invalid.",
    KeylessErrorCategory.INVALID_STATE,
    KeylessErrorResolutionTip.REAUTHENTICATE_UNSURE,
  ],
  [KeylessErrorType.INVALID_PROOF_VERIFICATION_KEY_NOT_FOUND]: [
    "The verification key used to authenticate was updated.",
    KeylessErrorCategory.SESSION_EXPIRED,
    KeylessErrorResolutionTip.REAUTHENTICATE,
  ],
  [KeylessErrorType.INVALID_JWT_SIG]: [
    "The JWK was found, but JWT failed verification",
    KeylessErrorCategory.INVALID_STATE,
    KeylessErrorResolutionTip.REAUTHENTICATE_UNSURE,
  ],
  [KeylessErrorType.INVALID_JWT_JWK_NOT_FOUND]: [
    "The JWK required to verify the JWT could not be found. The JWK may have been rotated out.",
    KeylessErrorCategory.SESSION_EXPIRED,
    KeylessErrorResolutionTip.REAUTHENTICATE,
  ],
  [KeylessErrorType.INVALID_JWT_ISS_NOT_RECOGNIZED]: [
    "The JWT issuer is not recognized.",
    KeylessErrorCategory.INVALID_STATE,
    KeylessErrorResolutionTip.UPDATE_REQUEST_PARAMS,
  ],
  [KeylessErrorType.INVALID_JWT_FEDERATED_ISS_NOT_SUPPORTED]: [
    "The JWT issuer is not supported by the Federated Keyless ",
    KeylessErrorCategory.API_ERROR,
    KeylessErrorResolutionTip.REAUTHENTICATE_UNSURE,
  ],
  [KeylessErrorType.INVALID_TW_SIG_VERIFICATION_FAILED]: [
    "The training wheels signature is invalid.",
    KeylessErrorCategory.INVALID_STATE,
    KeylessErrorResolutionTip.REAUTHENTICATE_UNSURE,
  ],
  [KeylessErrorType.INVALID_TW_SIG_PUBLIC_KEY_NOT_FOUND]: [
    "The public key used to verify the training wheels signature was not found.",
    KeylessErrorCategory.SESSION_EXPIRED,
    KeylessErrorResolutionTip.REAUTHENTICATE,
  ],
  [KeylessErrorType.INVALID_EXPIRY_HORIZON]: [
    "The expiry horizon is invalid.",
    KeylessErrorCategory.SESSION_EXPIRED,
    KeylessErrorResolutionTip.REAUTHENTICATE,
  ],
  [KeylessErrorType.JWK_FETCH_FAILED]: [
    "Failed to fetch JWKS.",
    KeylessErrorCategory.EXTERNAL_API_ERROR,
    KeylessErrorResolutionTip.JOIN_SUPPORT_GROUP,
  ],
  [KeylessErrorType.JWK_FETCH_FAILED_FEDERATED]: [
    "Failed to fetch JWKS for Federated Keyless provider.",
    KeylessErrorCategory.EXTERNAL_API_ERROR,
    KeylessErrorResolutionTip.JOIN_SUPPORT_GROUP,
  ],
  [KeylessErrorType.RATE_LIMIT_EXCEEDED]: [
    "Rate limit exceeded. Too many requests in a short period.",
    KeylessErrorCategory.API_ERROR,
    KeylessErrorResolutionTip.RATE_LIMIT_EXCEEDED,
  ],
  [KeylessErrorType.PEPPER_SERVICE_INTERNAL_ERROR]: [
    "Internal error from Pepper service.",
    KeylessErrorCategory.API_ERROR,
    KeylessErrorResolutionTip.SERVER_ERROR,
  ],
  [KeylessErrorType.PEPPER_SERVICE_BAD_REQUEST]: [
    "Bad request sent to Pepper service.",
    KeylessErrorCategory.API_ERROR,
    KeylessErrorResolutionTip.UPDATE_REQUEST_PARAMS,
  ],
  [KeylessErrorType.PEPPER_SERVICE_OTHER]: [
    "Unknown error from Pepper service.",
    KeylessErrorCategory.API_ERROR,
    KeylessErrorResolutionTip.SERVER_ERROR,
  ],
  [KeylessErrorType.PROVER_SERVICE_INTERNAL_ERROR]: [
    "Internal error from Prover service.",
    KeylessErrorCategory.API_ERROR,
    KeylessErrorResolutionTip.SERVER_ERROR,
  ],
  [KeylessErrorType.PROVER_SERVICE_BAD_REQUEST]: [
    "Bad request sent to Prover service.",
    KeylessErrorCategory.API_ERROR,
    KeylessErrorResolutionTip.UPDATE_REQUEST_PARAMS,
  ],
  [KeylessErrorType.PROVER_SERVICE_OTHER]: [
    "Unknown error from Prover service.",
    KeylessErrorCategory.API_ERROR,
    KeylessErrorResolutionTip.SERVER_ERROR,
  ],
  [KeylessErrorType.JWT_PARSING_ERROR]: [
    "Error when parsing JWT. This should never happen. Join https://t.me/+h5CN-W35yUFiYzkx for support",
    KeylessErrorCategory.INVALID_STATE,
    KeylessErrorResolutionTip.REINSTANTIATE,
  ],
  [KeylessErrorType.FULL_NODE_CONFIG_LOOKUP_ERROR]: [
    "Error when looking up on-chain keyless configuration.",
    KeylessErrorCategory.API_ERROR,
    KeylessErrorResolutionTip.SERVER_ERROR,
  ],
  [KeylessErrorType.FULL_NODE_VERIFICATION_KEY_LOOKUP_ERROR]: [
    "Error when looking up on-chain verification key.",
    KeylessErrorCategory.API_ERROR,
    KeylessErrorResolutionTip.SERVER_ERROR,
  ],
  [KeylessErrorType.FULL_NODE_JWKS_LOOKUP_ERROR]: [
    "Error when looking up on-chain JWKS.",
    KeylessErrorCategory.API_ERROR,
    KeylessErrorResolutionTip.SERVER_ERROR,
  ],
  [KeylessErrorType.FULL_NODE_OTHER]: [
    "Unknown error from full node.",
    KeylessErrorCategory.API_ERROR,
    KeylessErrorResolutionTip.SERVER_ERROR,
  ],
  [KeylessErrorType.UNKNOWN]: [
    "An unknown error has occurred.",
    KeylessErrorCategory.UNKNOWN,
    KeylessErrorResolutionTip.UNKNOWN,
  ],
};

export class KeylessError extends Error {
  readonly innerError?: unknown;

  readonly category: KeylessErrorCategory;

  readonly resolutionTip: KeylessErrorResolutionTip;

  readonly type: KeylessErrorType;

  readonly details?: string;

  /** @internal this constructor is for sdk internal use - do not instantiate outside of the SDK codebase */
  constructor(args: {
    innerError?: unknown;
    category: KeylessErrorCategory;
    resolutionTip: KeylessErrorResolutionTip;
    type: KeylessErrorType;
    message?: string;
    details?: string;
  }) {
    const { innerError, category, resolutionTip, type, message = KeylessErrors[type][0], details } = args;
    super(message);
    this.name = "KeylessError";
    this.innerError = innerError;
    this.category = category;
    this.resolutionTip = resolutionTip;
    this.type = type;
    this.details = details;
    this.message = KeylessError.constructMessage(message, resolutionTip, innerError, details);
  }

  static constructMessage(
    message: string,
    tip: KeylessErrorResolutionTip,
    innerError?: unknown,
    details?: string,
  ): string {
    let result = `\nMessage: ${message}`;
    if (details) {
      result += `\nDetails: ${details}`;
    }
    if (innerError instanceof AptosApiError) {
      result += `\nAptosApiError: ${innerError.message}`;
    } else if (innerError !== undefined) {
      result += `\nError: ${getErrorMessage(innerError)}`;
    }
    result += `\nKeylessErrorResolutionTip: ${tip}`;
    return result;
  }

  /**
   * Static constructor that creates a KeylessError instance using the KeylessErrors constant
   * @param args.type The type of KeylessError
   * @param args.aptosApiError optional AptosApiError supplied for api errors
   * @param args.details optional details to include in the error message
   * @returns A new KeylessError instance
   */
  static fromErrorType(args: { type: KeylessErrorType; error?: unknown; details?: string }): KeylessError {
    const { error, type, details } = args;

    const [message, category, resolutionTip] = KeylessErrors[type];
    return new KeylessError({
      message,
      details,
      innerError: error,
      category,
      resolutionTip,
      type,
    });
  }
}

/**
 * Options for handling errors in the Aptos API.
 */
type AptosApiErrorOpts = {
  apiType: AptosApiType;
  aptosRequest: AptosRequest;
  aptosResponse: AptosResponse<any, any>;
};

/**
 * Represents an error returned from the Aptos API.
 * This class encapsulates the details of the error, including the request URL, response status, and additional data.
 *
 * @param name - The name of the error, which is always "AptosApiError".
 * @param url - The URL to which the request was made.
 * @param status - The HTTP response status code (e.g., 400).
 * @param statusText - The message associated with the response status.
 * @param data - The response data returned from the API.
 * @param request - The original AptosRequest that triggered the error.
 */
export class AptosApiError extends Error {
  readonly url: string;

  readonly status: number;

  readonly statusText: string;

  readonly data: any;

  readonly request: AptosRequest;

  /**
   * Constructs an instance of AptosApiError with relevant error details.
   *
   * @param opts - The options for creating the AptosApiError.
   * @param opts.apiType - The type of API that generated the error.
   * @param opts.aptosRequest - The request object that caused the error.
   * @param opts.aptosResponse - The response object containing error details.
   *
   * @internal This constructor is for SDK internal use - do not instantiate outside the SDK codebase.
   */
  constructor({ apiType, aptosRequest, aptosResponse }: AptosApiErrorOpts) {
    super(deriveErrorMessage({ apiType, aptosRequest, aptosResponse }));

    this.name = "AptosApiError";
    this.url = aptosResponse.url;
    this.status = aptosResponse.status;
    this.statusText = aptosResponse.statusText;
    this.data = aptosResponse.data;
    this.request = aptosRequest;
  }
}

/**
 * Derives an error message from the Aptos API response, providing context for debugging.
 * This function helps in understanding the nature of the error encountered during an API request.
 *
 * @param {AptosApiErrorOpts} opts - The options for deriving the error message.
 * @param {AptosApiType} opts.apiType - The type of API being called.
 * @param {AptosRequest} opts.aptosRequest - The original request made to the Aptos API.
 * @param {AptosResponse} opts.aptosResponse - The response received from the Aptos API.
 */
function deriveErrorMessage({ apiType, aptosRequest, aptosResponse }: AptosApiErrorOpts): string {
  // eslint-disable-next-line max-len
  // extract the W3C trace_id from the response headers if it exists. Some services set this in the response, and it's useful for debugging.
  // See https://www.w3.org/TR/trace-context/#relationship-between-the-headers .
  const traceId = aptosResponse.headers?.traceparent?.split("-")[1];
  const traceIdString = traceId ? `(trace_id:${traceId}) ` : "";

  const errorPrelude: string = `Request to [${apiType}]: ${aptosRequest.method} ${
    aptosResponse.url ?? aptosRequest.url
  } ${traceIdString}failed with`;

  // handle graphql responses from indexer api and extract the error message of the first error
  if (apiType === AptosApiType.INDEXER && aptosResponse.data?.errors?.[0]?.message != null) {
    return `${errorPrelude}: ${aptosResponse.data.errors[0].message}`;
  }

  // Received well-known structured error response body - simply serialize and return it.
  // We don't need http status codes etc. in this case.
  if (aptosResponse.data?.message != null && aptosResponse.data?.error_code != null) {
    return `${errorPrelude}: ${JSON.stringify(aptosResponse.data)}`;
  }

  // This is the generic/catch-all case. We received some response from the API, but it doesn't appear to be a well-known structure.
  // We print http status codes and the response body (after some trimming),
  // in the hope that this gives enough context what went wrong without printing overly huge messages.
  return `${errorPrelude} status: ${aptosResponse.statusText}(code:${
    aptosResponse.status
  }) and response body: ${serializeAnyPayloadForErrorMessage(aptosResponse.data)}`;
}

const SERIALIZED_PAYLOAD_TRIM_TO_MAX_LENGTH = 400;

/**
 * This function accepts a payload of any type (probably an object) and serializes it to a string
 * Since we don't know the type or size of the payload, and we don't want to add a huge object in full to the error message
 * we limit the to the first 200 and last 200 characters of the serialized payload and put a "..." in the middle.
 * @param payload - The payload to serialize, which can be of any type.
 *
 * @returns A string representation of the serialized payload, potentially truncated.
 */
function serializeAnyPayloadForErrorMessage(payload: any): string {
  const serializedPayload = JSON.stringify(payload);
  if (serializedPayload.length <= SERIALIZED_PAYLOAD_TRIM_TO_MAX_LENGTH) {
    return serializedPayload;
  }
  return `truncated(original_size:${serializedPayload.length}): ${serializedPayload.slice(
    0,
    SERIALIZED_PAYLOAD_TRIM_TO_MAX_LENGTH / 2,
  )}...${serializedPayload.slice(-SERIALIZED_PAYLOAD_TRIM_TO_MAX_LENGTH / 2)}`;
}
