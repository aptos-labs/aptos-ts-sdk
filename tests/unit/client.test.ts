import { AptosApiError } from "../../src/errors";
import { AptosApiType } from "../../src/utils/const.js";

describe(AptosApiError.name, () => {
  it("should generate pretty error messages", () => {
    const err = new AptosApiError({
      apiType: AptosApiType.PEPPER,
      aptosRequest: {
        url: "http://blabla.com/my/api:8080",
        method: "POST",
      },
      aptosResponse: {
        data: { error: "something went wrong" },
        status: 400,
        statusText: "Bad Request",
        url: "http://blabla.com/my/api:8080",
        headers: {},
      },
    });

    expect(err.message).toMatchSnapshot();

    const err2 = new AptosApiError({
      apiType: AptosApiType.INDEXER,
      aptosRequest: {
        url: "http://blabla.com/my/api:8080",
        method: "POST",
      },
      aptosResponse: {
        data: {
          data: null,
          errors: [
            {
              message: "Your graphql query is invalid.",
            },
          ],
        },
        status: 200,
        statusText: "OK",
        url: "http://blabla.com/my/api:8080",
        headers: {},
      },
    });

    expect(err2.message).toMatchSnapshot();

    const err3 = new AptosApiError({
      apiType: AptosApiType.INDEXER,
      aptosRequest: {
        url: "http://blabla.com/my/api:8080",
        method: "POST",
      },
      aptosResponse: {
        data: {
          // some kinda large response payload
          foo: "bar",
          bar: "baz",
          nested: {
            asssdf: "asdf",
            asdf: ["asdfadsafds", "Asdfadafs"],
            assssdf: "asdf",
            afsdf: ["asdfadsafds", "Asdfadafs"],
            assassdf: "asdf",
            asasfdssdf: "asdf",
            asdff: ["asdfadsafds", "Asdfadafs"],
            assfsdf: "asdf",
            assfddf: ["asdfadsaasfdfds", "Asdfadafs"],
            asdfssdf: "asdf",
            asasdfdf: ["asdfadsafds", "Asdfadafs"],
            a2sdf: ["asdfadsafds", "Asdfadafs"],
            a2ssssdf: "asdf",
            ccasdfasdfsaasdf: ["asdfadsafds", "Asdfadafs"],
            cassssdf: "asdf",
            caasasfdssdf: "asdf",
            assdf: ["asdfadsafds", "Asdfadafs"],
            assafadsfsssdf: "asdf",
            asssasadfsfddf: ["asdfadsaasfdfds", "Asdfadafs"],
            asfdfssdf: "asdf",
            asaasdfsdfdf: ["asdfadsafds", "Asdfadafs"],
            afsasdfadfdf: ["asdfadsafds", "Asdfadafs"],
            assasdfsafdssdf: "asdf",
            asfdf: ["asdfadsafds", "Asdfadafs"],
            afadsfsssssdf: "asdf",
            aasfdafsasfdssdf: "asdf",
            asadsfasdfdf: ["asdfadsafds", "Asdfadafs"],
            assasdfadfssdf: "asdf",
            assadsfafdsfddf: ["asdfadsaasfdfds", "Asdfadafs"],
            asdadffasfssdf: "asdf",
            asdfss: ["asdfadsafds", "Asdfadafs"],
          },
        },
        status: 200,
        statusText: "OK",
        url: "http://blabla.com/my/api:8080",
        headers: {
          traceparent: "00-4312e4b208ab6749e42fc534e8590424-e94f817439ef9429-01",
        },
      },
    });
    expect(err3.message).toMatchSnapshot();
  });
});
