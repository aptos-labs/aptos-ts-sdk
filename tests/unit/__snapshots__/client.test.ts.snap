// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AptosApiError should generate pretty error messages 1`] = `"Request to [Pepper]: POST http://blabla.com/my/api:8080 failed with status: Bad Request(code:400) and response body: {"error":"something went wrong"}"`;

exports[`AptosApiError should generate pretty error messages 2`] = `"Request to [Indexer]: POST http://blabla.com/my/api:8080 failed with: Your graphql query is invalid."`;

exports[`AptosApiError should generate pretty error messages 3`] = `"Request to [Indexer]: POST http://blabla.com/my/api:8080 (trace_id:4312e4b208ab6749e42fc534e8590424) failed with status: OK(code:200) and response body: truncated(original_size:973): {"foo":"bar","bar":"baz","nested":{"asssdf":"asdf","asdf":["asdfadsafds","Asdfadafs"],"assssdf":"asdf","afsdf":["asdfadsafds","Asdfadafs"],"assassdf":"asdf","asasfdssdf":"asdf","asdff":["asdfadsafds",...dafsasfdssdf":"asdf","asadsfasdfdf":["asdfadsafds","Asdfadafs"],"assasdfadfssdf":"asdf","assadsfafdsfddf":["asdfadsaasfdfds","Asdfadafs"],"asdadffasfssdf":"asdf","asdfss":["asdfadsafds","Asdfadafs"]}}"`;
